x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -433,7 +461,7 @@ void searchShards(
                     skippedShards++;
                     continue;
                 }
-                List<DiscoveryNode> allocatedNodes = new ArrayList<>(group.allocatedNodes().size());
+                List<DiscoveryNode> allocatedNodes = Collections.synchronizedList(new ArrayList<>(group.allocatedNodes().size()));
```

Comment:
I wonder if we should make the logic that resolves the new target nodes a helper method and call it under sendingLock to avoid handling concurrency.